### ISSUE TITLE
Group & Sale Pricing Interoperability

### ISSUE THREAD

**Issue by joelbrock** [CORE-POS/IS4C#952]
Item has a groupprice of 6 @ 3.99. Item is on sale for .99 each.  While onsale the groupprice fails to kick in at 6 rings and the price for 6 remains at 5.94 (6*.99).  What would be involved in making  CORE-Lane capable of noticing this price collision and either using the group price, or prompting the cashier to select the price it should use?  Ideally the alert would appear in CORE-Office when staff is entering prices, or batch-building.

>Currently, when there is an item with a "bundled price" (i.e. 6-for-$3.99) that is also given a sales price (i.e. $0.99/each), the sales price will disable the bundled price, such that the customer fails to get the deal for the bundle. 
For example, a customer is charged $5.94 if they buy 6 Dr. Zevia's on sale for $0.99/each, rather than $3.99, which is the "bundled price" we've set at 6-for-$3.99. We need the bundled price to be prioritized above the sales price, such that when an item goes on sale, for example, for $0.99/each, the "bundled price" (6-for-$3.99) still remains active and is unaffected by the sales price.	

In the above example the item looks like (abridged):
```
                   upc: 0089477300115
           description: Zevia Dr. Zevia
                 brand: Zevia
          normal_price: 1.29
           pricemethod: 2
            groupprice: 3.99
              quantity: 6
         special_price: 0.99
    specialpricemethod: 0
     specialgroupprice: 0.99
       specialquantity: 0
         special_limit: 0
            start_date: 2017-12-13 00:00:00
              end_date: 2018-01-02 00:00:00
          mixmatchcode: 0
               batchID: 4311
              discount: 1
          discounttype: 1
          qttyEnforced: 1
```

**Comment by gohanman** [comment]
The issue here is mainly backward compatibility with a bit of UI complexity.

The "special" fields should override the "normal" fields - i.e.,

| On Sale Use | Not On Sale Use |
| ------------- | ------------------ |
| special_price | normal_price |
| specialpricemethod | pricemethod |
| specialgroupprice | groupprice |
| specialquantity | quantity |

I'm not gung-ho about breaking the clean division between these two. The interactions among four fields can get complicated enough without adding cross-over ways that all eight can interact.

The backward compatibility (and UI) issues in `batchList` which looks like this (omitting fields that aren't relevant):

| BatchList | ... |
| ---------- | --- |
| salePrice | been around forever |
| groupSalePrice | relatively new |
| pricemethod | maps to specialpricemethod |
| quantity | maps to specialquantity |

The UI issue is that by default `salePrice` and `groupSalePrice` contain the same value. I think that's mostly fine. There's an advanced batch editor that basically lets you directly edit the table. At a certain point just exposing the fields and documenting how to use them is going to work better than having a UI that covers every permutation and edge case intuitively.

The compatibility issue is a) not everyone has a `batches.groupSalePrice` column and b) both `batches.salesPrice` and `batches.groupSalePrice` might contain the same value that's intended to be used as a group price. In both cases `products.normal_price` is copied over to `products.special_price` to provide a price for partial group rings (this is strictly a backend consideration; the lane side only looks at the four "special" fields on a sale item).

What should probably happen is:
* If both `batches.salePrice` and `batches.groupSalePrice` and have different numeric values
  * Copy `batches.salePrice` to `products.special_price` and `batches.groupSalePrice` to `products.specialgroupprice`
* Otherwise use the old behavior
  * Copy `products.normal_price` to `products.special_price` and `batches.salePrice` to `products.specialgroupprice`

**Comment by joelbrock** [comment]
More end-user feedback:
>Their preference would simply be a checkbox in the item's group price settings that says "Override sales price?" and would allow the groupprice to override the sales price. 
So then, they items could still go in a sales batch for $0.99/ea, but the 6-for-$3.99 groupprice would override the individual sales price.

**Comment by joelbrock** [comment]
....and If I'm reading your remarks correctly, would the fix needed be a simple change to the SalesBatchTask?

**Comment by gohanman** [comment]
Sort of. There's still this part where the user has to drop down to effectively editing the underlying table to set all the necessary fields:

> The UI issue is that by default salePrice and groupSalePrice contain the same value. I think that's mostly fine. There's an advanced batch editor that basically lets you directly edit the table. At a certain point just exposing the fields and documenting how to use them is going to work better than having a UI that covers every permutation and edge case intuitively.

I strongly dislike the checkbox idea. It just adds complexity and scatters control over effective price more places. It simply can't work for one of the two options in the group price settings, and it's even in cases where it "works" I'm not sure how to handle all the new scenarios introduced. What happens if the 6-for-$3.99 group price is set to override sale price and someone puts it on sale for 6-for-$3.49? Instead of trying to guess what the user meant I'd much rather make the user be explicit about what should happen and have the decision(s) self-contained within the batch.

**Comment by gohanman** [comment]
Implemented in 2.9.0 with low-level tinkering of `batchList` records required.
